Fix application profile link mess

ccli8 opened the pull request:
### Summary of changes <!-- Required -->

This PR tries to address #407, including:
1.  Provide one approach to address cmake link mess with `mbed-core-flags`/
    `mbed-rtos-flags` whose links or not depends on selected application profile,
    `full` or `bare-metal`.
    1.  Add target config option `application-profile` for application to
        select which application profile to use. Applications are responsible
        for make this target config option consitent with link selection of
        mbed-os/mbed-baremetal.
    2.  Add cmake interface library `mbed-application-profile-flags`. It is to
        substitute for mbed-core-flags/mbed-rtos-flags. Library camke targets
        should change to link this one to build upon correct application profile
        selection.
2.  All in-tree library targets link `mbed-application-profile-flags` instead of `mbed-core-flags`/`mbed-rtos-flags` to build upon correct application profile.


#### Impact of changes <!-- Optional -->

To link mbed-baremetal, `target.application-profile` must be set to `bare-metal` in mbed_app.json5.

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Add an X to any of the following boxes that this PR functions as.
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------

multiplemonomials commented:
Woah, this is a lot to take on!  Thank you for working on this!

Haven't looked at the code yet but just going on the description:
- Could we rename `mbed-application-profile-flags` to either `mbed-os-flags` or `mbed-flags`?  Going to be typing that name 100000 times so I'd like it a bit shorter, and that doesn't lose information?
- We will need to provide an alias so that libraries and applications that use `mbed-core-flags` will still work.  Luckily CMake makes that easy to do with `add_library(ALIAS)`

ccli8 commented:
> Could we rename mbed-application-profile-flags to either mbed-os-flags or mbed-flags? Going to be typing that name 100000 times so I'd like it a bit shorter, and that doesn't lose information.

That's fine. Rename `mbed-application-profile-flags` to  `mbed-os-flags`.

> We will need to provide an alias so that libraries and applications that use mbed-core-flags will still work. Luckily CMake makes that easy to do with add_library(ALIAS)

There is compatible concern. My new idea is:
- new `mbed-core-flags` = old `mbed-core-flags`
- new `mbed-rtos-flags` = old `mbed-rtos-flags`
- added `mbed-os-flags` = old `mbed-core-flags` + old `mbed-rtos-flags` if `target.application-profile` is`full`.

To achieve above compatible, updated idea would become:
- new `mbed-core-flags` = old `mbed-core-flags` + old `mbed-rtos-flags` if `target.application-profile` is`full`.
- new `mbed-rtos-flags` = old `mbed-rtos-flags`
- added `mbed-os-flags` = new `mbed-core-flags` or become unnecessary

That is, `mbed-core-flags` is overloaded dependent on `target.application-profile`. And only one of `mbed-os` and `mbed-baremetal` target would be created dependent on `target.application-profile`, but not both. For example, when `target.application-profile` is `full`, `mbed-baremetal` cannot link `mbed-core-flags` because `mbed-core-flags` also links `mbed-rtos-flags`.
.

multiplemonomials commented:
Makes sense to me!  I definitely think that, depending on the application profile setting, we should make only mbed-baremetal or mbed-os available.

ccli8 commented:
Update as below:
1. Drop `mbed-application-profile-flags` (or its renamed versions `mbed-os-flags`). It is replaced with overloaded `mbed-core-flags`.
2. Overload `mbed-core-flags`/`mbed-core-sources` so that they will involve `mbed-rtos-flags`/`mbed-rtos-sources` or not, depending on `target.application-profile` setting.
3. All libraries must link `mbed-core-flags`  and no `mbed-rtos-flags`, or the link mess issue remains.
4. Both `mbed-os` and `mbed-baremetal` are still created, but one will become invalid and get caught on application link, depending on `target.application-profile` setting.

multiplemonomials commented on targets/targets.json5:
```diff
@@ -100,6 +100,10 @@
                 "value": false
+            },
+            "application-profile": {
+                "help": "Select application profile. Options: auto, full, bare-metal.",
```
What is the behavior of "auto"? Could we document that in the help here? Also, could we give a brief description of the difference between full and bare-metal? For example:

> full mode enables the real-time OS and enables all features of Mbed to be used. bare-metal mode disables the real-time OS, preventing you from using threads and libraries that depend on them. bare-metal mode is recommended for target MCUs with small amounts of flash and RAM.

ccli8 replied:
`auto` mode will check other build options and resolve to `full` mode by default or `bare-metal` mode, for example, `MBED_GREENTEA_TEST_BAREMETAL` being ON for Greentea test.

Updated. the help.

multiplemonomials replied:
Ah, we can remove the MBED_GREENTEA_TEST_BAREMETAL option.  It has never actually been used yet, sorry for the confusion!

ccli8 replied:
No, `MBED_GREENTEA_TEST_BAREMETAL` is used in CI:
https://github.com/mbed-ce/mbed-os/blob/25b05a10ec8e1ce25318817587bc77a9cca61418/tools/cmake/mbed_greentea.cmake#L95-L99

multiplemonomials replied:
Ah right whoops.  Well, I still think it might be easier to just use a JSON option only though.  Like, it uses a different mbed_app.json in CI for full vs baremetal, right? We could just stick the option in there and remove MBED_GREENTEA_TEST_BAREMETAL

ccli8 replied:
ok, I will keep the `auto` option and remove `MBED_GREENTEA_TEST_BAREMETAL` check.

ccli8 replied:
Update as below:
1. Keep the `auto` option and remove `MBED_GREENTEA_TEST_BAREMETAL` related check. Just reserve for future use.
2. For Greentea test build, check `target.application-profiile` setting to resolve link library instead of `MBED_GREENTEA_TEST_BAREMETAL`.

JohnK1987 commented:
In targets.json is also this parameter
https://github.com/mbed-ce/mbed-os/blob/722c2f119647e14f1f44d6e1e54041b4442e63b4/targets/targets.json5#L32-L33
I'm not sure if this has an use case at this moment.

ccli8 commented:
> In targets.json is also this parameter `supported_application_profiles`

The added `application-profile` config option and its supported options `full` and `bare-metal` just follows the naming of `supported_application_profiles`.

multiplemonomials commented on tools/cmake/mbed_target_functions.cmake:
```diff
@@ -136,9 +136,20 @@ function(mbed_set_post_build target)
+            get_target_property(MBED_CORE_FLAGS_TARGET_LINK_LIBRARIES mbed-core-flags INTERFACE_LINK_LIBRARIES)
             if("mbed-os" IN_LIST POST_BUILD_TARGET_LINK_LIBRARIES)
+                if(NOT "mbed-rtos-flags" IN_LIST MBED_CORE_FLAGS_TARGET_LINK_LIBRARIES)
+                    message(FATAL_ERROR
```
What situation would cause this error to trigger? How would `mbed-rtos-flags` end up as an interface link library of `mbed-core-flags`?

multiplemonomials replied:
Oh wait I see, this is basically checking the build profile.

multiplemonomials commented on CMakeLists.txt:
```diff
@@ -116,10 +116,28 @@ endif()
+                                         # depending on target.application-profile setting
+
+if("MBED_CONF_TARGET_APPLICATION_PROFILE=full" IN_LIST MBED_CONFIG_DEFINITIONS)
+    set(application_profile_config_full TRUE)
```
Nitpick: Could we make these uppercase to align with our usual CMake convention? Thanks!

ccli8 replied:
Uppercase these cmake variables

multiplemonomials commented on tools/cmake/mbed_target_functions.cmake:
```diff
@@ -136,9 +136,20 @@ function(mbed_set_post_build target)
             if("mbed-os" IN_LIST POST_BUILD_TARGET_LINK_LIBRARIES)
+                if(NOT "mbed-rtos-flags" IN_LIST MBED_CORE_FLAGS_TARGET_LINK_LIBRARIES)
+                    message(FATAL_ERROR
+                        "Miamatch of link libraries in Target ${target}: mbed-os without mbed-rtos-flags. Check target.application-profile setting."
```
Could we change this error message to say something like:
```
Target ${target} links to mbed-os, but Mbed is configured for a baremetal build. Please set "target.application-profile": "bare-metal" in JSON to enable the mbed-baremetal target.
```

ccli8 replied:
Updated

multiplemonomials commented on tools/cmake/mbed_target_functions.cmake:
```diff
@@ -136,9 +136,20 @@ function(mbed_set_post_build target)
             elseif("mbed-baremetal" IN_LIST POST_BUILD_TARGET_LINK_LIBRARIES)
+                if("mbed-rtos-flags" IN_LIST MBED_CORE_FLAGS_TARGET_LINK_LIBRARIES)
+                    message(FATAL_ERROR
+                        " Miamatch of link libraries in Target ${target}: mbed-baremetal with mbed-rtos-flags. Check target.application-profile setting."
```
Could we change this error message to say something like:
```
Target ${target} links to mbed-baremetal, but Mbed is configured for a fullbuild. Please set "target.application-profile": "bare-metal" in JSON to enable the mbed-baremetal target, or link your application to mbed-os instead of mbed-baremetal.
```

ccli8 replied:
As above

multiplemonomials reviewed (approved):
Thank you for making this fix!
